lib/docker-delta.coffee

```diff
@@ -104,69 +104,86 @@ parseDeltaStream = (input) ->
+		return null
+	.disposer(->)
+
+hardlinkCopy = (srcRoot, dstRoot, linkDests) ->
+	rsyncArgs = [
+		'--timeout', '300'
+		'--archive'
+		'--delete'
+	]
+	rsyncArgs.push('--link-dest', dest) for dest in linkDests
+	rsyncArgs.push(srcRoot, dstRoot)
+	rsync = spawn('rsync', rsyncArgs)
+	utils.waitPidAsync(rsync)
+
 exports.applyDelta = (srcImage) ->
 	deltaStream = new stream.PassThrough()
-
+	rootDirFunc = nullDisposer
 	if srcImage?
-		srcRoot = docker.imageRootDir(srcImage)
-		.then (srcRoot) ->
-			# trailing slashes are significant for rsync
-			srcRoot = path.join(srcRoot, '/')
-	else
-		srcRoot = null
+		rootDirFunc = docker.imageRootDirMounted.bind(docker)
 
-	dstId = parseDeltaStream(deltaStream).get('dockerConfig').bind(docker).then(docker.createEmptyImage)
+	dstIdPromise = parseDeltaStream(deltaStream).get('dockerConfig').bind(docker).then(docker.createEmptyImage)
 
-	Promise.all [
-		docker.infoAsync().get('Driver')
-		srcRoot
-		dstId
-		dstId.then(docker.imageRootDir)
-	]
-	.spread (dockerDriver, srcRoot, dstId, dstRoot) ->
-		# trailing slashes are significant for rsync
-		dstRoot = path.join(dstRoot, '/')
-
-		rsyncArgs = [
-			'--timeout', '300'
-			'--archive'
-			'--delete'
-			'--read-batch', '-'
-			dstRoot
-		]
+	Promise.using rootDirFunc(srcImage), (srcRoot) ->
+		srcRoot = path.join(srcRoot, '/') if srcRoot?
 
-		Promise.attempt ->
-			switch dockerDriver
-				when 'btrfs'
-					if srcRoot?
-						btrfs.deleteSubvolAsync(dstRoot)
-						.then ->
-							btrfs.snapshotSubvolAsync(srcRoot, dstRoot)
-				when 'overlay'
-					if srcRoot?
-						rsyncArgs.push('--link-dest', srcRoot)
+		Promise.all [
+			docker.infoAsync().get('Driver')
+			dstIdPromise
+			dstIdPromise.then(docker.imageRootDir)
+		]
+		.spread (dockerDriver, dstId, dstRoot) ->
+			# trailing slashes are significant for rsync
+			dstRoot = path.join(dstRoot, '/')
+
+			Promise.attempt ->
+				switch dockerDriver
+					when 'btrfs'
+						if srcRoot?
+							btrfs.deleteSubvolAsync(dstRoot)
+							.then ->
+								btrfs.snapshotSubvolAsync(srcRoot, dstRoot)
+					when 'overlay'
+						if srcRoot?
+							hardlinkCopy(srcRoot, dstRoot, [ srcRoot ])
+					when 'aufs'
+						if srcRoot?
+							docker.aufsDiffPaths(srcImage)
+							.then (diffPaths) ->
+								hardlinkCopy(srcRoot, dstRoot, diffPaths)
+					else
+						throw new Error("Unsupported driver #{dockerDriver}")
+			.then ->
+				rsyncArgs = [
+					'--timeout', '300'
+					'--archive'
+					'--delete'
+					'--read-batch', '-'
+					dstRoot
+				]
+				rsync = spawn('rsync', rsyncArgs)
+				deltaStream.pipe(rsync.stdin)
+
+				utils.waitPidAsync(rsync)
+			.then ->
+				# rsync doesn't fsync by itself
+				utils.waitPidAsync(spawn('sync'))
+			.then ->
+				deltaStream.emit('id', dstId)
+		.catch (e) ->
+			# If the process failed for whatever reason, cleanup the empty image
+			dstIdPromise.then (dstId) ->
+				docker.getImage(dstId).removeAsync()
+				.catch (e) ->
+					deltaStream.emit('error', e)
```

Comment:
That's a good point, this is the original way it was done but I'll change it so that it's in the correct order.